File: comparison/programs/mandelbrot.rb
def u8(value)
  value & 0xFF
end

def smul(a, b)
  sign = 0
  ma = a
  if a >= 128
    ma = u8(-a)
    sign += 1
  end
  mb = b
  if b >= 128
    mb = u8(-b)
    sign += 1
  end
  ah, al = ma.divmod(16)
  bh, bl = mb.divmod(16)
  result = u8(ah * bh * 16 + ah * bl + al * bh + al * bl / 16)
  sign == 1 ? u8(-result) : result
end

def mandel(cr, ci)
  zr = zi = iterations = 0
  loop do
    azr = zr >= 128 ? u8(-zr) : zr
    azi = zi >= 128 ? u8(-zi) : zi
    return iterations if azr > 32 || azi > 32

    zr2, zi2, zri = smul(zr, zr), smul(zi, zi), smul(zr, zi)
    zr, zi = u8(zr2 - zi2 + cr), u8(zri + zri + ci)
    iterations += 1
    return iterations if iterations >= 20
  end
end

ci = 236
while ci != 21
  line = +""
  cr = 216
  while cr != 25
    n = mandel(cr, ci)
    line << if n >= 20
              "#"
            elsif n >= 12
              "%"
            elsif n >= 8
              "+"
            elsif n >= 5
              ":"
            elsif n >= 3
              "."
            else
              " "
            end
    cr = u8(cr + 1)
  end
  puts line
  ci = u8(ci + 1)
end

File: comparison/programs/mandelbrot.go
package main

// Source for the generated mandelbrot.b Brainfuck program.
// Signed fixed-point multiplication uses scale 16 (16 units = 1.0).
func smul(a, b byte) byte {
	sign := byte(0)
	ma := a
	if a >= 128 {
		ma = 0 - a
		sign++
	}
	mb := b
	if b >= 128 {
		mb = 0 - b
		sign++
	}
	ah, al := ma/16, ma%16
	bh, bl := mb/16, mb%16
	result := ah*bh*16 + ah*bl + al*bh + al*bl/16
	if sign == 1 {
		result = 0 - result
	}
	return result
}

func mandel(cr, ci byte) byte {
	zr, zi, iterations := byte(0), byte(0), byte(0)
	done := byte(0)
	for done == 0 {
		azr := zr
		if zr >= 128 {
			azr = 0 - zr
		}
		azi := zi
		if zi >= 128 {
			azi = 0 - zi
		}
		if azr > 32 || azi > 32 {
			done = 1
		}
		if done == 0 {
			zr2, zi2, zri := smul(zr, zr), smul(zi, zi), smul(zr, zi)
			zr, zi = zr2-zi2+cr, zri+zri+ci
			iterations++
			if iterations >= 20 {
				done = 1
			}
		}
	}
	return iterations
}

func main() {
	for ci := byte(236); ci != 21; ci++ {
		for cr := byte(216); cr != 25; cr++ {
			n := mandel(cr, ci)
			if n >= 20 {
				print("#")
			} else if n >= 12 {
				print("%")
			} else if n >= 8 {
				print("+")
			} else if n >= 5 {
				print(":")
			} else if n >= 3 {
				print(".")
			} else {
				print(" ")
			}
		}
		print("\n")
	}
}
